product/api.go

```diff
@@ -0,0 +1,118 @@
+// dependencies map in the app.ProductManifest.
+//
+// The service shall be registered via app.PermissionKey service key.
+type PermissionService interface {
+	HasPermissionToTeam(userID, teamID string, permission *model.Permission) bool
+}
+
+// ClusterService enables to publish cluster events. In addition to that, It's being used for
+// mattermost-plugin-api Mutex API with the SetPluginKeyWithOptions method.
+//
+// The service shall be registered via app.ClusterKey key.
+type ClusterService interface {
+	PublishPluginClusterEvent(productID string, ev model.PluginClusterEvent, opts model.PluginClusterEventSendOptions) error
+	PublishWebSocketEvent(productID string, event string, payload map[string]interface{}, broadcast *model.WebsocketBroadcast)
+	SetPluginKeyWithOptions(productID string, key string, value []byte, options model.PluginKVSetOptions) (bool, *model.AppError)
+}
+
+// ChannelService provides channel related API  The service implementation is provided by
+// Channels product therefore the consumer products should add this service key to their
+// dependencies map in the app.ProductManifest.
+//
+// The service shall be registered via app.ChannelKey service key.
+type ChannelService interface {
+	GetDirectChannel(userID1, userID2 string) (*model.Channel, *model.AppError)
+	GetChannelByID(channelID string) (*model.Channel, *model.AppError)
+	GetChannelMember(channelID string, userID string) (*model.ChannelMember, *model.AppError)
+}
+
+// LicenseService provides license related utilities.
+//
+// The service shall be registered via app.LicenseKey service key.
+type LicenseService interface {
+	GetLicense() *model.License
+	RequestTrialLicense(requesterID string, users int, termsAccepted bool, receiveEmailsAccepted bool) *model.AppError
+}
+
+// UserService provides user related utilities. Initially this was thought to be app/users.UserService
+// but it's replaced by app.App temporarily. The reason is; UserService is a standalone tool whereas the
+// existing plugin API was using channels related app functionalities as well. We shall improve the UserService
+// to meet emerging requirements.
+//
+// The service shall be registered via app.UserKey service key.
+type UserService interface {
+	GetUser(userID string) (*model.User, *model.AppError)
+	UpdateUser(user *model.User, sendNotifications bool) (*model.User, *model.AppError)
+	GetUserByEmail(email string) (*model.User, *model.AppError)
+	GetUserByUsername(username string) (*model.User, *model.AppError)
+}
+
+// TeamService provides team related utilities.
+//
+// The service shall be registered via app.TeamKey service key.
+type TeamService interface {
+	GetMember(teamID, userID string) (*model.TeamMember, error)
+	CreateMember(ctx *request.Context, teamID, userID string) (*model.TeamMember, error)
+}
+
+// BotService is just a copy implementation of mattermost-plugin-api EnsureBot method.
```

Comment:
Let's also add another comment in the EnsureBot implementation that this has been copied over from https://github.com/mattermost/mattermost-plugin-api/blob/f1804b11e2b829d92474812d2f05875e49cd3777/bot.go#L147